File: platform/cli/src/commands/secret.ts
import { Command } from 'commander';
import chalk from 'chalk';
import ora from 'ora';
import * as readline from 'node:readline';
import * as fs from 'node:fs';
import * as api from '../api';
import { statusIcon, table, errorMessage } from '../format';

interface Secret {
  name: string;
  source?: string;
}

interface Backend {
  name: string;
  status: string;
  detail?: string;
}

export function registerSecretCommand(program: Command): void {
  const secret = program
    .command('secret')
    .description('Manage secrets');

  // Set a secret
  secret
    .command('set')
    .description('Set a secret value')
    .argument('<name>', 'Secret name')
    .argument('[value]', 'Secret value (reads from stdin if omitted)')
    .action(async (name: string, value?: string) => {
      try {
        if (value) {
          // Value provided on command line
          await api.put(`/api/secrets/${encodeURIComponent(name)}`, { value });
          console.log(`${chalk.green('\u2713')} Secret ${chalk.bold(name)} set`);
          return;
        }

        // Check if stdin has data (piped)
        if (!process.stdin.isTTY) {
          const chunks: Buffer[] = [];
          for await (const chunk of process.stdin) {
            chunks.push(chunk as Buffer);
          }
          const stdinValue = Buffer.concat(chunks).toString().trim();
          if (stdinValue) {
            await api.put(`/api/secrets/${encodeURIComponent(name)}`, {
              value: stdinValue,
            });
            console.log(`${chalk.green('\u2713')} Secret ${chalk.bold(name)} set`);
            return;
          }
        }

        // Interactive prompt with hidden input
        const rl = readline.createInterface({
          input: process.stdin,
          output: process.stdout,
        });

        const secretValue = await new Promise<string>((resolve) => {
          process.stdout.write(`Enter value for ${chalk.bold(name)}: `);
          const stdin = process.stdin;

          // Try to hide input
          if (stdin.isTTY) {
            stdin.setRawMode(true);
          }

          let input = '';
          const onData = (data: Buffer) => {
            const char = data.toString();
            if (char === '\n' || char === '\r') {
              stdin.removeListener('data', onData);
              if (stdin.isTTY) {
                stdin.setRawMode(false);
              }
              process.stdout.write('\n');
              rl.close();
              resolve(input);
            } else if (char === '\u007f' || char === '\b') {
              // Backspace
              if (input.length > 0) {
                input = input.slice(0, -1);
              }
            } else if (char === '\u0003') {
              // Ctrl+C
              process.stdout.write('\n');
              process.exit(0);
            } else {
              input += char;
            }
          };

          stdin.on('data', onData);
        });

        if (!secretValue) {
          console.error(chalk.red('No value provided'));
          process.exit(1);
        }

        await api.put(`/api/secrets/${encodeURIComponent(name)}`, {
          value: secretValue,
        });
        console.log(`${chalk.green('\u2713')} Secret ${chalk.bold(name)} set`);
      } catch (err) {
        console.error(errorMessage(err));
        process.exit(1);
      }
    });

  // List secrets
  secret
    .command('list')
    .description('List secret names')
    .action(async () => {
      try {
        const data = await api.get<Secret[]>('/api/secrets');

        if (!data || data.length === 0) {
          console.log(chalk.dim('No secrets configured.'));
          return;
        }

        for (const s of data) {
          if (s.source) {
            console.log(`${s.name}  ${chalk.dim(`(${s.source})`)}`);
          } else {
            console.log(s.name);
          }
        }
      } catch (err) {
        console.error(errorMessage(err));
        process.exit(1);
      }
    });

  // Delete a secret
  secret
    .command('delete')
    .description('Delete a secret')
    .argument('<name>', 'Secret name')
    .action(async (name: string) => {
      try {
        await api.del(`/api/secrets/${encodeURIComponent(name)}`);
        console.log(`${chalk.green('\u2713')} Secret ${chalk.bold(name)} deleted`);
      } catch (err) {
        console.error(errorMessage(err));
        process.exit(1);
      }
    });

  // Import from .env file
  secret
    .command('import')
    .description('Import secrets from a .env file')
    .argument('<file>', 'Path to .env file')
    .action(async (file: string) => {
      try {
        if (!fs.existsSync(file)) {
          console.error(chalk.red(`File not found: ${file}`));
          process.exit(1);
        }

        const content = fs.readFileSync(file, 'utf-8');
        const lines = content.split('\n');
        let count = 0;

        const spinner = ora('Importing secrets...').start();

        for (const line of lines) {
          const trimmed = line.trim();
          // Skip empty lines and comments
          if (!trimmed || trimmed.startsWith('#')) continue;

          const eqIdx = trimmed.indexOf('=');
          if (eqIdx === -1) continue;

          const key = trimmed.slice(0, eqIdx).trim();
          let val = trimmed.slice(eqIdx + 1).trim();

          // Strip surrounding quotes
          if (
            (val.startsWith('"') && val.endsWith('"')) ||
            (val.startsWith("'") && val.endsWith("'"))
          ) {
            val = val.slice(1, -1);
          }

          if (key) {
            await api.put(`/api/secrets/${encodeURIComponent(key)}`, {
              value: val,
            });
            count++;
          }
        }

        spinner.succeed(`Imported ${count} secret${count !== 1 ? 's' : ''} from ${file}`);
      } catch (err) {
        console.error(errorMessage(err));
        process.exit(1);
      }
    });

  // Sync secrets from cloud backends
  secret
    .command('sync')
    .description('Sync secrets from cloud backends')
    .action(async () => {
      try {
        const spinner = ora('Syncing secrets from backends...').start();
        const result = await api.post<{ synced: number }>('/api/secrets/sync');
        spinner.succeed(
          `Synced ${result?.synced ?? 0} secret${(result?.synced ?? 0) !== 1 ? 's' : ''} from backends`,
        );
      } catch (err) {
        console.error(errorMessage(err));
        process.exit(1);
      }
    });

  // Backend status
  secret
    .command('status')
    .description('Show backend status')
    .action(async () => {
      try {
        const data = await api.get<Backend[]>('/api/secrets/backends');

        if (!data || data.length === 0) {
          console.log(chalk.dim('No backends configured.'));
          return;
        }

        const rows = data.map((b) => {
          const icon = statusIcon(b.status);
          const detail = b.detail ? chalk.dim(` (${b.detail})`) : '';
          return [b.name, `${icon} ${b.status}${detail}`];
        });

        console.log(table(rows, ['Backend', 'Status']));
      } catch (err) {
        console.error(errorMessage(err));
        process.exit(1);
      }
    });
}

File: platform/cli-go/internal/commands/secret.go
package commands

import (
	"bufio"
	"fmt"
	"os"
	"strings"
	"time"

	"github.com/briandowns/spinner"
	"github.com/elloloop/devpod/platform/cli-go/internal/api"
	"github.com/elloloop/devpod/platform/cli-go/internal/format"
	"github.com/spf13/cobra"
	"golang.org/x/term"
)

type secretEntry struct {
	Name   string `json:"name"`
	Source string `json:"source"`
}

type backendEntry struct {
	Name   string `json:"name"`
	Status string `json:"status"`
	Detail string `json:"detail"`
}

func newSecretCmd() *cobra.Command {
	secretCmd := &cobra.Command{
		Use:   "secret",
		Short: "Manage secrets",
	}

	// secret set <name> [value]
	setCmd := &cobra.Command{
		Use:   "set <name> [value]",
		Short: "Set a secret value",
		Args:  cobra.RangeArgs(1, 2),
		RunE: func(cmd *cobra.Command, args []string) error {
			name := args[0]

			if len(args) > 1 {
				// Value provided on command line
				value := args[1]
				if err := api.Put("/api/secrets/"+name, map[string]string{"value": value}, nil); err != nil {
					return fmt.Errorf("%s", format.ErrorMsg(err.Error()))
				}
				fmt.Printf("%s Secret %s set\n", format.SuccessMsg("\u2713"), name)
				return nil
			}

			// Check if stdin has data (piped)
			stdinInfo, _ := os.Stdin.Stat()
			if stdinInfo.Mode()&os.ModeCharDevice == 0 {
				// Piped input
				scanner := bufio.NewScanner(os.Stdin)
				var lines []string
				for scanner.Scan() {
					lines = append(lines, scanner.Text())
				}
				stdinValue := strings.TrimSpace(strings.Join(lines, "\n"))
				if stdinValue != "" {
					if err := api.Put("/api/secrets/"+name, map[string]string{"value": stdinValue}, nil); err != nil {
						return fmt.Errorf("%s", format.ErrorMsg(err.Error()))
					}
					fmt.Printf("%s Secret %s set\n", format.SuccessMsg("\u2713"), name)
					return nil
				}
			}

			// Interactive prompt with hidden input
			fmt.Printf("Enter value for %s: ", name)
			password, err := term.ReadPassword(int(os.Stdin.Fd()))
			fmt.Println()
			if err != nil {
				return fmt.Errorf("%s", format.ErrorMsg("Failed to read input"))
			}

			secretValue := strings.TrimSpace(string(password))
			if secretValue == "" {
				return fmt.Errorf("%s", format.ErrorMsg("No value provided"))
			}

			if err := api.Put("/api/secrets/"+name, map[string]string{"value": secretValue}, nil); err != nil {
				return fmt.Errorf("%s", format.ErrorMsg(err.Error()))
			}
			fmt.Printf("%s Secret %s set\n", format.SuccessMsg("\u2713"), name)
			return nil
		},
	}

	// secret list
	listCmd := &cobra.Command{
		Use:   "list",
		Short: "List secret names",
		RunE: func(cmd *cobra.Command, args []string) error {
			var data []secretEntry
			if err := api.Get("/api/secrets", &data); err != nil {
				return fmt.Errorf("%s", format.ErrorMsg(err.Error()))
			}

			if len(data) == 0 {
				fmt.Println(format.DimText("No secrets configured."))
				return nil
			}

			for _, s := range data {
				if s.Source != "" {
					fmt.Printf("%s  %s\n", s.Name, format.DimText(fmt.Sprintf("(%s)", s.Source)))
				} else {
					fmt.Println(s.Name)
				}
			}
			return nil
		},
	}

	// secret delete <name>
	deleteCmd := &cobra.Command{
		Use:   "delete <name>",
		Short: "Delete a secret",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			name := args[0]
			if err := api.Delete("/api/secrets/" + name); err != nil {
				return fmt.Errorf("%s", format.ErrorMsg(err.Error()))
			}
			fmt.Printf("%s Secret %s deleted\n", format.SuccessMsg("\u2713"), name)
			return nil
		},
	}

	// secret import <file>
	importCmd := &cobra.Command{
		Use:   "import <file>",
		Short: "Import secrets from a .env file",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			file := args[0]

			if _, err := os.Stat(file); os.IsNotExist(err) {
				return fmt.Errorf("%s", format.ErrorMsg(fmt.Sprintf("File not found: %s", file)))
			}

			content, err := os.ReadFile(file)
			if err != nil {
				return fmt.Errorf("%s", format.ErrorMsg(err.Error()))
			}

			lines := strings.Split(string(content), "\n")
			count := 0

			s := spinner.New(spinner.CharSets[14], 100*time.Millisecond)
			s.Suffix = " Importing secrets..."
			s.Start()

			for _, line := range lines {
				trimmed := strings.TrimSpace(line)
				if trimmed == "" || strings.HasPrefix(trimmed, "#") {
					continue
				}

				eqIdx := strings.Index(trimmed, "=")
				if eqIdx == -1 {
					continue
				}

				key := strings.TrimSpace(trimmed[:eqIdx])
				val := strings.TrimSpace(trimmed[eqIdx+1:])

				// Strip surrounding quotes
				if len(val) >= 2 {
					if (val[0] == '"' && val[len(val)-1] == '"') || (val[0] == '\'' && val[len(val)-1] == '\'') {
						val = val[1 : len(val)-1]
					}
				}

				if key != "" {
					_ = api.Put("/api/secrets/"+key, map[string]string{"value": val}, nil)
					count++
				}
			}

			s.Stop()
			plural := "s"
			if count == 1 {
				plural = ""
			}
			fmt.Printf("%s Imported %d secret%s from %s\n", format.SuccessMsg("\u2713"), count, plural, file)
			return nil
		},
	}

	// secret sync
	syncCmd := &cobra.Command{
		Use:   "sync",
		Short: "Sync secrets from cloud backends",
		RunE: func(cmd *cobra.Command, args []string) error {
			s := spinner.New(spinner.CharSets[14], 100*time.Millisecond)
			s.Suffix = " Syncing secrets from backends..."
			s.Start()

			var result struct {
				Synced int `json:"synced"`
			}
			err := api.Post("/api/secrets/sync", nil, &result)
			s.Stop()
			if err != nil {
				return fmt.Errorf("%s", format.ErrorMsg(err.Error()))
			}
			plural := "s"
			if result.Synced == 1 {
				plural = ""
			}
			fmt.Printf("%s Synced %d secret%s from backends\n", format.SuccessMsg("\u2713"), result.Synced, plural)
			return nil
		},
	}

	// secret status
	statusCmd := &cobra.Command{
		Use:   "status",
		Short: "Show backend status",
		RunE: func(cmd *cobra.Command, args []string) error {
			var data []backendEntry
			if err := api.Get("/api/secrets/backends", &data); err != nil {
				return fmt.Errorf("%s", format.ErrorMsg(err.Error()))
			}

			if len(data) == 0 {
				fmt.Println(format.DimText("No backends configured."))
				return nil
			}

			var rows [][]string
			for _, b := range data {
				icon := format.StatusIcon(b.Status)
				detail := ""
				if b.Detail != "" {
					detail = format.DimText(fmt.Sprintf(" (%s)", b.Detail))
				}
				rows = append(rows, []string{b.Name, fmt.Sprintf("%s %s%s", icon, b.Status, detail)})
			}

			fmt.Println(format.Table([]string{"Backend", "Status"}, rows))
			return nil
		},
	}

	secretCmd.AddCommand(setCmd)
	secretCmd.AddCommand(listCmd)
	secretCmd.AddCommand(deleteCmd)
	secretCmd.AddCommand(importCmd)
	secretCmd.AddCommand(syncCmd)
	secretCmd.AddCommand(statusCmd)

	return secretCmd
}
